pumpkindb_engine/src/script/mod.rs

```diff
@@ -315,42 +315,65 @@ impl<'a, T: Dispatcher<'a>> Scheduler<'a, T> {
     /// Once an environment execution has been terminated, a message will be sent,
     /// depending on the result (`EnvTerminated` or `EnvFailed`)
     pub fn run(&mut self) {
+        let mut rng = thread_rng();
         let mut envs: VecDeque<(EnvId, Env<'a>, Sender<ResponseMessage>)> = VecDeque::new();
-
+        // Flag that indicates that the Env being processed should be removed from
+        // the queue
+        let mut pop_front = false;
+        // Queue length is dynamically updated to avoid scanning the queue
+        // every time we need to know its size
+        let mut len = 0;
         loop {
-            match envs.pop_front() {
-                Some((pid, mut env, chan)) => {
-                    let program = env.program[env.program.len() - 1];
-                    match self.pass(&mut env, pid.clone()) {
-                        Err(Error::Reschedule) => {
-                            env.program.push(program);
-                            envs.push_back((pid, env, chan));
-                        }
-                        Err(err) => {
-                            self.dispatcher.done(&mut env, pid);
-                            let stack_size = env.stack_size;
-                            let _ = chan.send(ResponseMessage::EnvFailed(pid,
-                                                                         err,
-                                                                         Some(env.stack_copy()),
-                                                                         Some(stack_size)));
-                        }
-                        Ok(()) => {
-                            if env.program.is_empty() ||
-                                (env.program.len() == 1 && env.program[0].len() == 0) {
-                                self.dispatcher.done(&mut env, pid);
+            {
+                // Borrow the front of the queue mutably
+                match envs.front_mut() {
+                    Some(&mut (pid, ref mut env, ref chan)) => {
```

Comment:
may help readability to give names and explicit structure to these tuples at some point